src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory_Methods.cs

```diff
@@ -44,153 +46,70 @@ private static ILiteralExpression CreateIncrementOneLiteralExpression(BoundIncre
             });
         }
 
-        private static readonly ConditionalWeakTable<BoundExpression, IArgument> s_argumentMappings = new ConditionalWeakTable<BoundExpression, IArgument>();
+        private static readonly ConditionalWeakTable<BoundExpression, IEnumerable<IArgument>> s_callToArgumentsMappings
+            = new ConditionalWeakTable<BoundExpression, IEnumerable<IArgument>>();
 
-        private static IArgument DeriveArgument(
-            int parameterIndex,
-            int argumentIndex,
+        internal static IArgument CreateArgumentOperation(ArgumentKind kind, IParameterSymbol parameter, IOperation value)
```

Comment:
@genlu I think at the end, everything related to IOperation should move into here.